File: 프로그래머스/2/131704. 택배상자/택배상자.py
def solution(order):
    answer = 0
    box = []
    second_box = []
    
    for i in range(len(order), 0, -1):
        box.append(i)
    
    idx = 0
    while box or second_box:
        now = order[idx]
        if box and box[-1] == now:
            box.pop()
            idx += 1
            answer += 1
        elif second_box and second_box[-1] == now:
            second_box.pop()
            idx += 1
            answer += 1
        elif box:
            second_box.append(box.pop())
        else:
            break
    
    return answer

File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        Queue<Integer> belt = new ArrayDeque<>();
        Stack<Integer> container = new Stack<>();
        for (int i = 1; i <= order.length; i++)
            belt.offer(i);
        
        int answer = 0;
        int idx = 0;
        while (!belt.isEmpty() || !container.isEmpty()) {
            int box = order[idx];
            int beltBox = !belt.isEmpty() ? belt.peek() : 0;
            int conBox = !container.isEmpty() ? container.peek() : 0;
            if (box == beltBox)
                belt.poll();
            else if (box == conBox)
                container.pop();
            else {
                if (belt.isEmpty())
                    break;
                else {
                container.push(belt.poll());
                continue;
                }
            }
            answer++;
            idx++;
        }
        
        return answer;
    }
}